### DISCUSSION TITLE
How to build this repo locally

### DISCUSSION THREAD

**Discussion by e-t-k** [CBIIT/ccdi-federation-api#20] [Announcements]
Hi all,
As a new Rust user I had to figure out the minimal steps to get this code running on my local machine (MacBook Air). I figure some of you may be in the same boat so I wanted to share my notes on the process. Any comments & suggestions are gratefully welcomed as I may be missing essential aspects.

Updated Feb 19, 2025 with anatomical site info and new folder structure.

**Set up the repo**
- [Install Rust](https://www.rust-lang.org/tools/install). This will get you the tools `rustc` (compiler), `rustup` (updater), and `cargo` (package management).
- Switch to the "nightly build": (I started with the stable build, but when I tried to compile the api I got an error in a dependency.)
  -  `rustup toolchain install nightly` to install
  - `rustup default nightly` to use
- Clone the repo: `git clone https://github.com/CBIIT/ccdi-federation-api.git`
- Go into the rust crates dir: `cd ccdi-federation-api/crates`. (This is the working directory for all following commands.)
- Check out your desired branch, eg: `git checkout feat/restrict-charsets`

**Run the server locally**
- To get the reference server running locally, run `cargo run --bin ccdi-spec serve`. This will build all the dependencies (takes a minute the first time and is quicker thereafter) and then start a server on http://localhost:8000/ . When I ran it, I got a warning popup:
> Do you want the application “cargo” to accept incoming network connections?
> Clicking Deny may limit the application’s behavior. This setting can be changed in the Firewall pane of Security & Privacy preferences.

I chose **Deny** and it seemed to work fine, since I'm connecting only from my local browser.

- To visit the swagger spec in the browser, navigate to http://localhost:8000/swagger-ui/ (the trailing slash is required!)
- You can also visit an API implementation with fake data by going to the appropriate endpoints, eg http://localhost:8000/sample/by/tumor_classification/count .

**Make changes**
- Edit the rust files in the [ccdi-spec](https://github.com/CBIIT/ccdi-federation-api/tree/main/crates/ccdi-spec) dir and descendants.
-  Don't edit the [swagger.yml](https://github.com/CBIIT/ccdi-federation-api/blob/main/swagger.yml) file directly.
- When you've made your edits, go into the `ccdi-spec` subdir and update the swagger file: `cargo run generate > ../../swagger.yml` . This will build your package and emit the YAML code. 
- To view the changed spec you will need to interrupt the server (eg Ctrl-C) and start it up again with `cargo run serve` as well as refresh the page in the browser.
- The default swagger compilation omits many anatomical sites. This is probably what you want while testing.
- When doing a final commit, include the anatomical sites by running the following, which takes several minutes: `cargo run --features all-anatomical-site generate > ../../swagger.yml`

**Comment by claymcleod** [comment]
This is great!

I also recommend these commands to check that everything is passing as it should _before_ you commit. You can create an alias for them if you want in your shell of choice—I just store them in my shell history and rerun them there.

```bash
# Check that all linting and formatting passes.
cd packages && cargo fmt && cargo test && cargo doc && cargo clippy -- -D warnings; cd ..

# Generate a new version of the spec, make sure it lints okay, then copy to
# pasteboard (on Mac) so that you can view it in editor.swagger.io.
#
# Note: you can also just run the server and visit the specification at 
# http://localhost:8000/swagger-ui/ as Ellen points out above.
cd packages && cargo run generate -o ../swagger.yml -f && cd .. && spectral lint swagger.yml && cat swagger.yml | pbcopy
```

If you want to run the latter command, you'll need to install Node (`brew install node npm` [with Homebrew](brew.sh) on Mac) and then install the `spectral` tool (`npm install -g @stoplight/spectral-cli`).

**Comment by e-t-k** [comment]
As of November 8,  there is a new package, `ccdi-curate`, which downloads uberon and compiles the rust class.
Clay demoed it with the following example commands:

```
cargo run --release
cargo run --release --bin ccdi-curate
cargo run --release --bin ccdi-curate uberon -vv
```

**Comment by e-t-k** [comment]
### Validate endpoints vs reference
You can validate your endpoints vs the reference implementation to confirm that they conform.
The command to do this is `check`:
`cargo run --bin ccdi-spec check <URL> <RESPONSE_TYPE>`

Examples:
`cargo run --bin ccdi-spec check "https://ccdi.treehouse.gi.ucsc.edu/api/v1/info" Information`

`cargo run --bin ccdi-spec check "https://ccdi.treehouse.gi.ucsc.edu/api/v1/sample?page=2&diagnosis=9380/3 : Glioma, malignant" Samples`

If the bottom line is `Success!`, the response was conformant (this may not catch every error).
Otherwise, you will get an error. For example:
-  `error: data did not match any variant of untagged enum Description at line 1 column 329`.
- `error: missing field `counts` at line 10 column 1`

This gives you a hint of where in the JSON your endpoint response diverged from what the reference implementation was expecting. 

Possible RESPONSE_TYPEs, also listed when you call `cargo run --bin ccdi-spec check --help` :
`Samples, Sample, SamplesByCount, Subjects, Subject, SubjectsByCount, Files, Namespaces, Namespace, Organizations, Organization, Summary, Information, FieldDescriptions, Errors`
